DICOM/src/Microsoft.Health.Anonymizer.Common/CryptoHashFunction.cs

```diff
@@ -0,0 +1,101 @@
+{
+    public class CryptoHashFunction
+    {
+        private readonly HMAC _hmac;
+
+        public CryptoHashFunction(CryptoHashSetting cryptoHashSetting)
+        {
+            EnsureArg.IsNotNull(cryptoHashSetting, nameof(cryptoHashSetting));
+
+            byte[] byteKey = cryptoHashSetting.GetCryptoHashByteKey();
+            _hmac = cryptoHashSetting.CryptoHashType switch
+            {
+                HashAlgorithmType.Md5 => new HMACMD5(byteKey),
+                HashAlgorithmType.Sha1 => new HMACSHA1(byteKey),
+                HashAlgorithmType.Sha256 => new HMACSHA256(byteKey),
+                HashAlgorithmType.Sha512 => new HMACSHA512(byteKey),
+                HashAlgorithmType.Sha384 => new HMACSHA384(byteKey),
+                _ => throw new AnonymizerException(AnonymizerErrorCode.CryptoHashFailed, "Hash function not supported."),
+            };
+        }
+
+        public byte[] Hash(byte[] input)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+
+            return Hash(input, _hmac);
+        }
+
+        public byte[] Hash(Stream input)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+
+            return Hash(input, _hmac);
+        }
+
+        public byte[] Hash(string input, Encoding encoding = null)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+
+            return Hash(input, _hmac, encoding);
+        }
+
+        public FixedLengthString Hash(FixedLengthString input, Encoding encoding = null)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+
+            return Hash(input, _hmac, encoding);
+        }
+
+        public static byte[] Hash(byte[] input, HMAC hashAlgorithm)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+            EnsureArg.IsNotNull(hashAlgorithm, nameof(hashAlgorithm));
+
+            return hashAlgorithm.ComputeHash(input);
+        }
+
+        public static byte[] Hash(string input, HMAC hashAlgorithm, Encoding encoding = null)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+            EnsureArg.IsNotNull(hashAlgorithm, nameof(hashAlgorithm));
+
+            encoding ??= Encoding.UTF8;
+            return hashAlgorithm.ComputeHash(encoding.GetBytes(input));
+        }
+
+        public static byte[] Hash(Stream input, HMAC hashAlgorithm)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+            EnsureArg.IsNotNull(hashAlgorithm, nameof(hashAlgorithm));
+
+            return hashAlgorithm.ComputeHash(input);
+        }
+
+        public static FixedLengthString Hash(FixedLengthString input, HMAC hashAlgorithm, Encoding encoding = null)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+            EnsureArg.IsNotNull(hashAlgorithm, nameof(hashAlgorithm));
+
+            var hashData = Hash(input.ToString(), hashAlgorithm, encoding);
+
+            input.SetString(string.Concat(hashData.Select(b => b.ToString("x2"))));
```

Comment:
🟡 SUGGESTION
Not sure if it's meaningful to perform hash on Fixed Length String. Because hash results are usually fixed length, if you trim or pad the hash result to your origin length, the hash transformation seems useless here. Do we have any scenarios for this feature?